herbert_core/utilities/geometry/box_intersect.m

```diff
@@ -0,0 +1,156 @@
+function inter_points = box_intersect(box_minmax,cross_plain)
+% Calculate intersection points between the box in ND (2,3,4)
+% and line/plain/hyperplane of dimension (N-1)D
+%
+% Inputs:
+% box_minmax -- NDx2 array of min and max points of the box, to intersect with.
+%
+% cross_plain -- NDxND array of points defining plain in the appropriate
+%                dimensions. The coordinates defined along the first
+%                dimension and the second dimention correspont to number of
+%                points, the plain is passing through (has to be ND). If
+%                there are ND-1 points, the missing point assumed to be
+%                equal to 0.
+% Outputs:
+% inter_points - NDxNp where NP -- the number of intesection points
+%                array of points, defining intersection between the
+%                edges of the box and the line/plain/hyperplain defined as
+%                the second argument
+%                If no intersection points are present, the array is empty
+
+ndim = size(cross_plain,1);
+switch(ndim)
+    case(2)
+        inter_points = intersect2D(box_minmax,cross_plain);
+    case(3)
+        inter_points = intersect3D(box_minmax,cross_plain);
+    case(4)
+        inter_points = intersect4D(box_minmax,cross_plain);
+    otherwise
+        error('BOX_INTERSECT:invalid_argument',...
+            'Routine accepts the data from 2 to 4 dimensions. Got %d',...
+            ndim);
+end
+function inter_points = intersect4D(box_minmax,cross_plain)
```

Comment:
I would keep the switching for now. While Alex has provided a generic extensible method with, as you say, similar code, it may be that in future we would want to do simpler implementations for the lower-dimension geometries and this enables us to keep that separation.

Reply:
there is 50% code duplication here. Reducing it would mean defining generic normal and indeed building NDintersect. Though I theoretically know how to do it now ( Levi-Civita symbol in ND) doing so is a bit of an overkill. I am not sure if I really need 4D. Currently the requested improvement does not worth the time spent to do it.